### ISSUE TITLE
script for cleaning multiple rooms

### ISSUE THREAD

**Issue by Kijan** [humbertogontijo/homeassistant-roborock#252]
hi

I read this:
Action for making a roborock vacuum a specific room configured in the official app? #141
I got it working to clean rooms and so I added a script for each room.
Is it possible to put a helper in any way in the place where the room number has to go?
I would like to select rooms on my floorplan and with a delay or another button i would like to start the cleaning of the selected rooms.

**Comment by CH1PSET1** [comment]
Also interested in this moment. I would like to be able to select certain rooms for cleaning.

**Comment by Romario1972** [comment]
> I got it working to clean rooms and so I added a script for each room.

Could you please share your script for one room please?

**Comment by Kijan** [comment]
i created a script for each room like this:
`alias: Büro saugen
sequence:
  - service: roborock.vacuum_clean_segment
    data:
      segments:
        - 20
    target:
      entity_id: vacuum.sawi
  - service: input_boolean.turn_on
    data: {}
    target:
      entity_id: input_boolean.vacuum_buro
mode: single
icon: mdi:robot-vacuum
`

**Comment by Kijan** [comment]
yeah but this is not flexibel.
i have an interface with a button for each room.
i would like to press all buttons for the rooms i like to clean and then it triggers, but i am not so deep in the coding to get the segments flexible

**Comment by blindguynar** [comment]
Although this says it's for the Xiaomi Mii integration it works perfectly fine for this integration as well and allows for cleaning multi-rooms.   It uses queuing so the script runs once for each room in sequence.  I still set-up a script for each room that in-turn calls the larger queued script though so i could easily see the buttons i pressed.  Although there are alternate ways to do this as well also.  

Only minor updates to the script were required (device id's, entity names, room names, etc) to make it work as it isn't specific to any integration. You create a number helper for each room (named specifically) and set it to the room segment id.  then call this script with that room name.  then call it again for the next room and it will queue those calls and execute them in-order until complete.  The runtime likely doesn't service a HA restart though as a note.

https://community.home-assistant.io/t/howto-buttons-for-room-cleaning-with-queuing-for-xiaomi-miio-robot-vacuum/320755

Hope this helps you out.

**Comment by Kijan** [comment]
thx i will try it ;)

**Comment by Klakurka** [comment]
> You can enter multiple segments to the service call:
> 
> ```
> sequence:
>   - service: roborock.vacuum_clean_segment
>     data:
>       segments:
>         - 16
>         - 17
>         - 18
>         - 19
>       repeats: 2
>     target:
>       entity_id: vacuum.roborock_s5_max
> ```

When I try this, it wakes up and leaves the dock but as soon as it backs out it just says it's done cleaning and goes back to dock.

Any ideas?

This is an S7 MaxV.

```
action:
  - service: vacuum.send_command
    data:
      command: app_segment_clean
      params:
        - segments:
            - 16
            - 17
        - repeat: 1
    target:
      entity_id: vacuum.cleaner
mode: single
```

**Comment by Lash-L** [comment]
@Klakurka make sure those segment ids are correct for you. And potentially remove the dash before repeats

**Comment by blindguynar** [comment]
Sorry i set it up awhile back, but it appears i changed the actual vac command to a service specific to this integration to originally get it working.  I just tested this simplified version of my script and it cleaned my kitchen without issue.  Please give it a shot.

in this case my segment id for kitchen is 16 and therefore i have a number helper called roborock_room_enum_kitchen that is set to the value 16 and i've set the script up to just have room kitchen and call the script for room kitchen.  

```
alias: Roborock Room Cleaning
max: 30
fields:
  room:
    name: Room
    required: true
    example: kitchen
    selector:
      select:
        options:
          - kitchen
sequence:
  - choose:
      - conditions:
          - condition: or
            conditions:
              - condition: state
                entity_id: vacuum.roborock_s7_maxv
                state: cleaning
              - condition: state
                entity_id: vacuum.roborock_s7_maxv
                state: error
        sequence:
          - wait_for_trigger:
              - platform: state
                entity_id: vacuum.roborock_s7_maxv
                to: returning
              - platform: state
                entity_id: vacuum.roborock_s7_maxv
                to: docked
            continue_on_timeout: false
            timeout: "01:30:00"
          - service: roborock.vacuum_clean_segment
            data:
              segments: >
                {% set room_id = states("input_number.roborock_room_enum_" +
                room) | int %} {{ room_id }}
            target:
              entity_id: vacuum.roborock_s7_maxv
          - wait_for_trigger:
              - platform: state
                entity_id: vacuum.roborock_s7_maxv
                to: returning
              - platform: state
                entity_id: vacuum.roborock_s7_maxv
                to: docked
            continue_on_timeout: false
            timeout: "01:30:00"
      - conditions:
          - condition: or
            conditions:
              - condition: state
                entity_id: vacuum.roborock_s7_maxv
                state: returning
              - condition: state
                entity_id: vacuum.roborock_s7_maxv
                state: docked
              - condition: state
                entity_id: vacuum.roborock_s7_maxv
                state: idle
        sequence:
          - service: roborock.vacuum_clean_segment
            data:
              segments: >
                {% set room_id = states("input_number.roborock_room_enum_" +
                room) | int %} {{ room_id }}
            target:
              entity_id: vacuum.roborock_s7_maxv
          - wait_for_trigger:
              - platform: state
                entity_id: vacuum.roborock_s7_maxv
                to: returning
              - platform: state
                entity_id: vacuum.roborock_s7_maxv
                to: docked
            continue_on_timeout: false
            timeout: "01:30:00"
    default: []
```

**Comment by Kijan** [comment]
> > yeah but this is not flexibel
> 
> It won't be more flexible, than being able to declare multiple rooms for the service call.

no if write the rooms directly in the script, it is not flexible at all. 
i cant choose the rooms on the dashboard.

**Comment by blindguynar** [comment]
> You can make a script which adding the rooms to the service call according to specific conditions. No need to create a script for every combination...
> 
> The problem with these script is, if you restart/reload HA the robot won't continue the cleaning, because HA sends the command to clean the next room after the robot finished with the current room. Also you can only stop the cleaning by aborting the script. For example, if you press the home button on the robot, it will start going to the dock, but HA will immediately send it to the next room.

Specific script per room?  Why would you do that?  I have one script for my entire top floor.  I call that one script and send it which room i want to clean.  Then call the same script again for the next room and it queues them.  There is just a SINGLE script so i don't know why you'd create more than one.  

Example of my main floor script showing many rooms within the same script.
```
  room:
    name: Room
    required: true
    example: kitchen
    selector:
      select:
        options:
          - kitchen
          - dining_room
          - master_bedroom
          - living_room
          - hall
          - laundry
          - office
          - guest_bedroom
          - hall_bathroom
sequence:
```

Return to Dock: That is just a button helper attached to an automation to stop the script, send the robot to the dock, and start the script.  It's a single button push and a simple automation.  From the app you're right.  Though i have an automation that will return the robot to the dock when it runs into issues too.  Basically i've not ran into the need to cancel it form the app verses just using a tablet dashboard.

i already mentioned it wouldn't survive a restart of HA so i can't solve that point.  In that case it will finish the room it's on and return to dock.  It will pick up next time on schedule so it's not like it's left in the middle of the room.  I don't find this as a huge deal for me and certainly it doesn't happen on the daily.  Your mileage certainly seems to vary from mine.

It seems to boil down to our definition of flexibility.  I don't consider predefined routes as flexible.  You don't and that's perfectly fine and meets your use case.  If you don't like the script you don't have to use it, but OP can also decide that for themselves as well.  You can certainly also write an even better one that also handles all the use cases, but with predefined segment lists it won't be quite as flexible under my definition of flexibility.

Basically OP had an issue open for 4 months for a specific requirement of flexibility, without a resolution.  I had a working solution that handles 95% of their use case so i thought i'd offer it to them.  Nothing more and nothing less.  Don't use it.  No one is forcing you to use it.

**Comment by blindguynar** [comment]
> > no if write the rooms directly in the script, it is not flexible at all. i cant choose the rooms on the dashboard.
> 
> Nobody said that you should hardcode anything in a script. Where did you read it? If you want to to choose the rooms on the dashboard, write the script so that you can. :D
> 
> The service call supports a dictionary for selected the rooms, that is all you need.
> 
> I have input_booleans for every room, and the cleaning script checks which of them is turned on, and makes the service call accordingly.

That is certainly one way to do it.  Of course you've got to flip them on/off before and after the script runs in order to create differing paths.  Additionally you'd need to wait until your current run is finished before starting the next run unless you implement queuing as this script does.  Here you can start additional rooms without interfering with the currently running rooms.  So i can add rooms in real-time to the path which you couldn't do if you called them for all segments from the beginning.

Can't believe a simple little script is causing this much of a uprising!  Queuing has its place guys.  Yes you can do this in other ways.  This is just one and it does have a few drawbacks, but also a few advantages too.

**Comment by blindguynar** [comment]
> Again... Where do you read that?? Because I'm sure did not write as such. I have a single script, which sends a single service call to the robot with the rooms it needs to clean. After that call the script is finished, and the robot cleans the selected room without HA supervision, just like when you do it from the Roborock app.
> 

I might have misquoted you.  Sorry.  Seemingly i read a comment about needing to create a script for each room.  Again Sorry.

The input booleans will/do certainly work as you've stated, but after you've made that service call and you'd like to add a room to the path how does your solution account for that?  Maybe it does and i just don't understand it correctly.

**Comment by Kijan** [comment]
> 

As i said i am not so deep in the scripts and your example here has multiple rooms at once. so if i understand that right, it would clean all the rooms at once. so this example would not be flexible. that is all i said

**Comment by Kijan** [comment]
> You can enter multiple segments to the service call:
> 
> ```
> sequence:
>   - service: roborock.vacuum_clean_segment
>     data:
>       segments:
>         - 16
>         - 17
>         - 18
>         - 19
>       repeats: 2
>     target:
>       entity_id: vacuum.roborock_s5_max
> ```

As i said i am not so deep in the scripts and your example here has multiple rooms at once. so if i understand that right, it would clean all the rooms at once. so this example would not be flexible. that is all i said

**Comment by blindguynar** [comment]
For dashboards i highly recommend the "Xiaomi Vacuum Map Card" for on-demand multi-room cleaning.  This queuing works well for automatic cleaning for me based on my scheduling not by time of day, but rather by time patterns along with room/home occupancy.  This is because my vac can be currently cleaning room A and my conditions also say Room B should be cleaned.  Room B is added to the queue without changing the current cleaning of room A.  

One example is its cleaning the kitchen as it does this most often (kids are messy!), but my current location has increased to  > 10 miles away from my home.  Then it adds the living room, dining room, & bathroom as additional rooms to clean knowing it will have the extra time to do it.  I have other automations that track how many times per day specific rooms are cleaned so they don't continue to fire if they've already fired enough times for that day/week and are reset at midnight for the day and friday for the week. 

It might be that my use case is on the edge, but for me it works well.

**Comment by blindguynar** [comment]
> Using input_booleans is just one way, you can use a selector as well. For me I have a wall mounted touch screen, and the easiest way to select the rooms, is a button for each room, which controlling an input_boolean. I only do manual triggered cleaning, no schedule, etc.
> 

Can i ask why you aren't just using the "Xiaomi Vacuum Map Card" for selecting rooms from a tablet?  If you aren't doing any scheduling and all of your cleaning is on-demand you don't need a booleans or a fancy script at all right?

Seemingly the Vacuum card is a much richer UI than simply selecting boolean toggles.  My home has multiple floors and this is my lower floor tablet (i have one on each floor).  By utilizing the multi map loading service and browser_mod.navigate i'm able to automatically display the correct map for each floor.  (The load map button isn't used anymore as i automated the load of the correct map per floor)

This is my lower floor.  I can select multi-rooms from the map and run them all at once.  No booleans or queue scripts at all are required to select multi room cleaning on-demand and it's absolutely a beautiful UI.  I use this for on-demand and queuing for scheduling.  
![IMG_1483](https://github.com/humbertogontijo/homeassistant-roborock/assets/48634544/fe0e1c5d-1cce-4f10-9f2a-16fb4e892895)

>
If you want queueing, that is very easy as well, change the script mode to queued, and add a wait trigger after the service call, so the script will be in running state while the robot is cleaning.
> 

How is that any different than this script?  It will have the exact same behavior.  Additionally my vacuum does NOT actually return to the dock when done with a room, it's instead re-tasked to the next room.  As soon as it's status is set to "returning" it tasks to the next room so it doesn't ever get back to the dock unless it needs to empty it's bin or clean it's mop.  If finishes a room and goes right to the next room.

Yes you could do that in a queued automation too.  Yes is does talk as it's being retasked to a room, but your queued automation would have the exact same behavior so i don't see the advantage.  

Additionally you can NOT start automations, but rather they have to be triggered.  You can start scripts anytime you want.  Yes you can tigger your automation off a toggle boolean, but that isn't as clean IMO as you have a nearly useless helper laying around as opposed to not needing it at all when using a script instead.

I'm actually not trying to convince you to use this script.  I'll sleep really good knowing my solution fits my use case exactly, but seemingly you're telling the OP it's not a good way and i do disagree with that advice.  It might be specific to the use case, but this method absolutely has "some" advantages over a boolean driven script for scheduling.  Specifically not needing to manually select booleans, dealing with cleaning them up after each run before running the next, and having extra input toggles just only exist to trigger an automation.

It seems we'll need to agree to disagree on implementation of our respective use cases.  I would be interested in why you would choose the toggle method over the map card if all you do is on-demand cleaning from a tablet as that seems like a less feature rich solution.

**Comment by blindguynar** [comment]
I understand the NSpanel angle now and why toggles would be better for your use case.  Sorry you said "touch display" in the reply so i wasn't thinking NSpanel.  I use a few NSpanels myself in some areas.

As far as restarts, I guess i might just be unique as i don't restart HA that often.  And the way my automations are setup for scheduling it wouldn't even matter if i did.  As i said all mine are based on time patterns and so they'll attempt to fire again on the next time pattern.  So if it gets interrupted it will just finish the current room, return to the dock, and fire again at the next time pattern match with zero intervention from me.  I'd never even know, or care, it happened as i wouldn't be around to see/hear it and it would just work the next time the dynamic schedule requested it.  

You seem to be under the impression i don't understand the cleaning segment command or its benefit.  I do understand that if i send it rooms 16, 17, and 18 all at once it will clean them all one after another.  And if HA restarts it would finish cleaning all three segments.  This is not in question as i assure i do certainly get that point.  And i get that "if" you know you want to clean all three at this exact moment it is a better way to do it all in a single command for sure.  

Our definitions of on-demand and dynamic seem to be at odds though since there is a major difference between what we are doing.  There is a major difference between on-demand (from a touch screen) telling the robot to clean a list of rooms right now vs. dynamically scheduling it to clean automatically based on a sophisticated set of dynamically changing conditions in real-time. 

i ONLY visit that tablet for a spill and otherwise I never have to click anything.  Our use cases could not be further from equal if they tried to be north/south poles of a magnet.  i am not doing on-demand cleaning such as you from a touch screen as my primary mode of cleaning and therefore our reasoning for using one method or another is never gonna equal.  

To put it another way is that when the automations start room 16 (no interaction from me) it does NOT know at that exact time that it can also clean room 17 and/or 18 and/or 22 and/or 25, and/or etc.  It might only end up being able to clean 16 or based on factors that occur 5 mins after room 16 begins it determines room 17 can also be cleaned before it believes i'll return home.  I'm not walking up to a touch screen to determine that behavior.  The automations are and i don't have to manually touch anything.  Its dynamically scheduling itself with no intervention from me based on room/home occupancy, calendar appointments, and my current direction of travel and distance from home.  In addition to all the other members of my family as well.  

There is a huge difference between that and selecting a list of rooms on a screen.

Could my automations be smarter and determine up-front that it could clean 16 and 17 as to reduce the # of service calls?  Most certainly and without debate, but this is a simpler form to gain the same implementation of function since i'm tracking 3 people's current locations among a vast array of other inputs/conditions as well. 

And your example script has the "exact" same talking behavior as a single room.  It will talk after running your segments.  So if you run 16,17,18 then queue up 20 & 21 it will talk in between the switch between the first and 2nd set.  So it talks less, but you should understand i'm just not around to hear the talking anyway so it doesn't matter to me.  The only one here is the dog when it cleans and she hasn't mentioned to me it bothers her.  

Only on a rare occasion do i even ever return home as the robot hasn't made it back to the dock.  My entire home is cleaned 2-3 times a week with certain rooms being cleaned upwards of 8 times a week.  i never see it, hear, know, or care it happened.  I just go to work, bed, dinner, store, lowe's, coffee, etc and if it can clean one room it does or 3 rooms or none or the entire floor.   Once it knows i'm headed home it raps up and if i get to close while it's in the middle of a room it cancels itself and returns to the dock.  I don't do a thing.  

i hope you'll see that this is quite a bit different than pre-selecting rooms on a screen.  For on-demand cleaning in which you tell the robot to clean a list of rooms upfront this is not an ideal solution and without question your solution of a single command to all known rooms is FAR and AWAY superior.  I don't question or debate that on single bit.  That is NOT my use case and i didn't believe it was the OP's use case either, but i now believe based on the comments from the OP they're not looking for a solution as in-depth as mine either.

For dynamically scheduling rooms in a highly sophisticated manner in which the system is determining a room cleaning schedule in a real-time changing environment without human intervention this solution works really well.  For an on-demand user interaction your method of a single call is superior.  

Again though our use cases could not be further mis-aligned if they were cops and robbers.

I don't think we're ever gonna be on the same page on this.  So again we're trying to do just two totally separate things.  if the rooms you want to clean are known upfront your single call is superior, but that is not what i'm doing as i do NOT know the rooms upfront.  The system decides the rooms and schedules dynamically for my family without intervention from us.
